farm-boosted-rewards-subscriber/src/service.rs

```diff
@@ -33,51 +34,73 @@ pub trait ServiceModule:
     + only_admin::OnlyAdminModule
 {
     #[endpoint(subtractPayment)]
-    fn subtract_payment_endpoint(
-        &self,
-        service_index: usize,
-        user_ids: MultiValueEncoded<AddressId>,
-    ) {
+    fn subtract_payment_endpoint(&self, user_ids: MultiValueEncoded<AddressId>) {
         self.require_caller_is_admin();
-        require!(
-            service_index == SubscriptionUserType::Normal as usize
-                || service_index == SubscriptionUserType::Premium as usize,
-            "Invalid service index"
-        );
-        let is_premium_service = service_index == SubscriptionUserType::Premium as usize;
+        let current_epoch = self.blockchain().get_block_epoch();
+        let payment_recurrency = EPOCHS_IN_WEEK;
+        let standard_service_index = SubscriptionUserType::Normal as usize;
+        let premium_service_index = SubscriptionUserType::Premium as usize;
         let energy_threshold = self.energy_threshold().get();
         let fees_contract_address = self.fees_contract_address().get();
-        let mut processed_user_ids = ManagedVec::new();
+        let mut standard_processed_user_ids = ManagedVec::new();
+        let mut premium_processed_user_ids = ManagedVec::new();
 
         for user_id in user_ids {
-            if is_premium_service {
-                let opt_user_address = self
-                    .user_id()
-                    .get_address_at_address(&fees_contract_address, user_id);
-                if opt_user_address.is_none() {
-                    continue;
-                }
-                let user = opt_user_address.unwrap();
-                let user_energy = self.get_energy_amount(&user);
-
-                if user_energy < energy_threshold {
-                    continue;
-                }
+            let opt_user_address = self
+                .user_id()
+                .get_address_at_address(&fees_contract_address, user_id);
+            if opt_user_address.is_none() {
+                continue;
             }
 
-            self.subtract_user_payment(fees_contract_address.clone(), service_index, user_id);
-            processed_user_ids.push(user_id);
+            let user_last_payment_mapper = self.user_last_payment(user_id);
+            let mut user_last_payment = if user_last_payment_mapper.is_empty() {
+                UserLastPayment::default()
+            } else {
+                user_last_payment_mapper.get()
+            };
+
+            if current_epoch < user_last_payment.epoch + payment_recurrency
+                && user_last_payment.epoch > 0
+            {
+                continue;
+            }
+
+            let user = opt_user_address.unwrap();
```

Comment:
```suggestion
            let user = unsafe { opt_user_address.unwrap_unchecked() };
```
You already checked for `None`.

Reply:
Done.